fish-rust/src/builtins/random.rs

```diff
@@ -0,0 +1,193 @@
+    }
+    fn parse<T: PrimInt>(streams: &mut io_streams_t, cmd: &wstr, num : &wstr) -> Result<T, wutil::Error> {
+        let res = fish_wcstoi_radix_all(num.chars(), None, true);
+        if res.is_err() {
+            streams.err.append(wgettext_fmt!(
+                "%ls: %ls: invalid integer\n",
+                cmd,
+                num,
+            ));
+        }
+        return res;
+    }
+
+    match arg_count {
+        0 => {
+            // Keep the defaults
+        },
+        1 => {
+            // Seed the engine persistently
+            let num = parse::<i64>(streams, cmd, argv[i]);
+            match num {
+                Err(_) => return STATUS_INVALID_ARGS,
+                Ok(x) => *engine = SmallRng::seed_from_u64(x as u64)
+            }
+            return STATUS_CMD_OK;
+        },
+        2 => {
+            // start is first, end is second
+            match parse::<i64>(streams, cmd, argv[i]) {
+                Err(_) => return STATUS_INVALID_ARGS,
+                Ok(x) => start = x
+            }
+
+            match parse::<i64>(streams, cmd, argv[i + 1]) {
+                Err(_) => return STATUS_INVALID_ARGS,
+                Ok(x) => end = x
+            }
+        },
+        3 => {
+            // start, step, end
+            match parse::<i64>(streams, cmd, argv[i]) {
+                Err(_) => return STATUS_INVALID_ARGS,
+                Ok(x) => start = x
+            }
+
+            // start, step, end
+            match parse::<u64>(streams, cmd, argv[i + 1]) {
+                Err(_) => return STATUS_INVALID_ARGS,
+                Ok(0) => {
+                    streams.err.append(wgettext_fmt!(
+                        "%ls: STEP must be a positive integer\n",
+                        cmd,
+                    ));
+                    return STATUS_INVALID_ARGS;
+                }
+                Ok(x) => step = x
+            }
+
+            match parse::<i64>(streams, cmd, argv[i + 2]) {
+                Err(_) => return STATUS_INVALID_ARGS,
+                Ok(x) => end = x
+            }
+        },
+        _ => {
+            streams.err.append(wgettext_fmt!(
+                "%ls: too many arguments\n",
+                cmd,
+            ));
+            return Some(1);
+        }
+    }
+
+    if end <= start {
+        streams.err.append(wgettext_fmt!(
+            "%ls: END must be greater than START\n",
+            cmd,
+        ));
+        return STATUS_INVALID_ARGS;
+    }
+
+    // Possibilities can be abs(i64::MIN) + i64::MAX,
+    // so we do this as i128
+    let possibilities = (end as i128 - start as i128) / (step as i128);
+
+    if possibilities == 0 {
+        streams.err.append(wgettext_fmt!(
+            "%ls: range contains only one possible value\n",
+            cmd,
+        ));
+        return STATUS_INVALID_ARGS;
+    }
+
+    let rand = engine.gen_range(0..=possibilities);
+
+    let result = start as i128 + rand as i128 * step as i128;
```

Comment:
```suggestion
    let result = start.checked_add_unsigned(rand * step).unwrap();
```

Reply:
Uuhhhh... why are we specifically doing a checked_add only to throw away the error?

Reply:
There's no unchecked `add_unsigned`, and this can't overflow.

Reply:
Ah, so the possibilities are positive, so rand and step are positive, so adding it to start mathematically has to fit.

So you're doing checked_add with unwrap here to crash if it doesn't?

Reply:
The problem is that `rand * step` is a `u64` that's quite possibly larger than `i64::MAX`, so you can't simply convert it to `i64` and add it. There's no implementation for `i64 + u64` at all, so we have to roll our own; alternatively we could also use `overflowing_add_unsigned` (overflow can never happen with these values anyway, so there's no difference in functionality).

Reply:
We could convert to u64, do the computation (with wrapping behavior), and convert the result back to i64.
Probably that's what the C++ version did. The result should be correct because it's two's complement.